pkg/gateway/controller.go

```diff
@@ -0,0 +1,245 @@
+		gatewayAvailable: true,
+	}
+	controller.clusterStatusMap.Store(make(map[string]bool))
+	return controller
+}
+
+func getNewClientsetFunc() NewClientsetFunc {
+	if NewClientset != nil {
+		return NewClientset
+	}
+	return dynamic.NewForConfig
+}
+
+func (c *Controller) Start(kubeConfig *rest.Config) error {
+	gwClientset, err := c.getCheckedClientset(kubeConfig)
+	if errors.IsNotFound(err) {
+		klog.Infof("gateways resource not found, disabling Gateway status controller")
+		c.gatewayAvailable = false
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	klog.Infof("Starting Gateway status Controller")
+	c.store, c.informer = cache.NewInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return gwClientset.List(metav1.ListOptions{})
+		},
+		WatchFunc: gwClientset.Watch,
+	}, &unstructured.Unstructured{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(obj interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			klog.V(2).Infof("GatewayStatus %q updated", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			klog.V(2).Infof("GatewayStatus %q deleted", key)
+			if err == nil {
+				c.gatewayDeleted(obj, key)
+			}
+		},
+	})
+	go c.informer.Run(c.stopCh)
+	go c.runWorker()
+
+	return nil
+}
+
+func (c *Controller) Stop() {
+	if c.gatewayAvailable {
+		close(c.stopCh)
+		c.queue.ShutDown()
+		klog.Infof("Gateway status Controller stopped")
+	}
+}
+
+func (c *Controller) runWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for Gateways stopped")
+			return
+		}
+
+		key := keyObj.(string)
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.store.GetByKey(key)
+			if err != nil {
+				klog.Errorf("Error retrieving gateway with key %q from the cache: %v", key, err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+				return
+			}
+			if exists {
+				c.gatewayCreatedOrUpdated(obj.(*unstructured.Unstructured))
+			}
+			c.queue.Forget(key)
+		}()
+	}
+}
+
+func (c *Controller) gatewayDeleted(obj interface{}, key string) {
+	var resource *unstructured.Unstructured
+	var ok bool
+	if resource, ok = obj.(*unstructured.Unstructured); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Could not convert object %v to DeletedFinalStateUnknown", obj)
+			return
+		}
+
+		resource, ok = tombstone.Obj.(*unstructured.Unstructured)
+		if !ok {
+			klog.Errorf("Could not convert object tombstone %v to Unstructured", tombstone.Obj)
+			return
+		}
+	}
+
+	haStatus, _, _ := getGatewayStatus(resource)
+	if haStatus == "active" {
+		c.clusterStatusMap.Store(make(map[string]bool))
+	}
```

Comment:
I think there's an edge case where this may be problematic. The normal case is where the active engine shuts down gracefully and deletes its `Gateway`.  However if it is killed or the delete fails then the `Gateway` will be stale. In this case, the engine has a stale `Gateway` check and thus the new engine will delete it later. This means we could get a deleted `Gateway` that has `haStatus == "active"` but isn't really the active `Gateway`. To handle this, we can keep track of the last `Gateway` resource that was created/updated in an `atomic.Value` via its UUID.  Here, if the the UUID matches the cached UUID then clear the map.

Reply:
Is it possible for stale gateway and new gateway to co-exist? as in, two gateways with `haStatus == "active"` ? If not, then it will eventually get overwritten and we should be fine. This logic will likely change once we get Active-ACctive HA.

Reply:
> Is it possible for stale gateway and new gateway to co-exist? as in, two gateways with `haStatus == "active"` ? 
 
yes - new engine instance will publish its new active `Gateway` while the stale active `Gateway` will remain until the engine purges it.

If not, then it will eventually get overwritten and we should be fine. This logic will likely change once we get Active-Active HA.
> 
> EDIT: Also, we get updates every 5 seconds, so we will end up filling it with correct information.

True but there may be a 5 sec gap where no clusters are connected and user requests will fail. Unless I'm missing something...

Reply:
Thinking some more, I'm wondering if we should do anything here, ie keep the last known state. Doesn't the engine leave the tunnels and network state as is on shutdown so as to maintain connectivity?  @sridhargaddam What do you think? Should we clear the cluster connectivity state or leave the last state when the `Gateway` is deleted, ie when the engine fails over.

Reply:
More I think about it, I like the idea of UUID. We will need to use it when we get Active-Active, so might as well get a head start on it.

Reply:
I think it would be more stable to cache the last active gateway update connection list and use that to calculate which clusters are up or down. 

You could have a deleted/recreated gateway on a failover, or you could have duplicated active, while that's not true (at least yet)

Also consider that during those events it could be healthy to still resolve to what we knew with the hope that reconnection will happen quickly as it was before. So you allow the resolution, and then the connection will eventually succeed.

It's different to the case where a connection to a cluster goes down. (Please note that a failover on a remote cluster can also trigger this on other clusters)

I wonder if it would be helpful to (at some sprint) implement a filter in your algorithms to ignore spurious transitions like this (<x seconds) .

Reply:
> Thinking some more, I'm wondering if we should do anything here, ie keep the last known state. Doesn't the engine leave the tunnels and network state as is on shutdown so as to maintain connectivity? @sridhargaddam What do you think? 

I'm working on preserving the tunnels (so that datapath is not disrupted) even when GatewayPod is restarted. So far could not get it working, as the pluto/charon process is getting killed as soon as GatewayPod is terminated. Looking into various options on how to get these processes running even when GatewayPod is terminated.

>Should we clear the cluster connectivity state or leave the last state when the `Gateway` is deleted, ie when the engine fails over.

Currently, as soon as GatewayPod is deleted, it WILL cleanup the Gateway object. **This is now properly handled with some recent fixes like clean termination of GatewayPod.** 
If we are able to successfully preserve the tunnels (even after the GatewayPod is terminated), then I think we should NOT delete the Gateway object anymore. Stale Gateway objects (based on the active leader election) should be handled either in route-agentPod or the new GatewayPod leader that is scheduled. This is a good topic to be discussed with the team.

@vthapar @tpantelis @mangelajo 
For now, I agree we can cache the last active gateway and use that to calculate which is up or down.

JYFI, it looks like we have bug in our implementation where connection status is not changing even when remoteGateway node is down.
https://github.com/submariner-io/submariner/issues/740